src/modm/platform/usb/stm32/usb.hpp.in

```diff
@@ -0,0 +1,49 @@
+/*
```

Comment:
I strongly disagree with making this a template just so you can code generate a name suffix for it, C++ templates was a much better choice because all the tooling works with it (IDE, clang-format, compiler errors, etc

I would prefer to keep the code generation to a minimum and use C++ as far as possible.

Reply:
I understand the problem and in this case we can probably elide code gen, however, the Peripheral is called differently across the STM32 portfolio (Usb, UsbOtgFs, UsbOtgHs), and has different CMSIS register names. I'd like to get tinyusb to work on all STM32 devices in the future, so I assume I'll have to configure some more stuff than just this.

Reply:
Ack, what primarily would make my life easier is if the API was not codegen (but impl files are OK). The reason for this is that the IDE auto-complete only really needs the header file to work reliably. It also allows the IDE to show the documentation etc on hover which is convenient.

So if we had:

* **.hpp** never templated
* **_impl.hpp / .cpp** possibly templated

...then I would happy in all cases except when I actually needed to work on the module (which is << the number of times I use it as library).

In your case, you could address this by moving the implementation of `UsbUart::initialize` into a `.cpp` file.

Reply:
You should only use the code completion on the generated output, not on the repository. It simply won't work for many reasons more than just the code gen. In fact one of the reasons for using code gen was to better support code completion, since the output is much smaller and much simpler to integrate into an IDE, because everything in the local folder is valid.

I agree that there is some overuse of lbuild when a CPP/C++ solution would do too. However, inside the `modm:platform` modules it's simply necessary to use codegen to preserve sanity. In addition it makes documentation so much easier, since you do not need to forward the exact macros also to Doxygen. Even then, things like configuring the buffer size for UART via lbuild options are things I would like to change to a pure C++ solution. But I don't have the time for it.

I don't see an alternative to using codegen this extensively, especially not with supporting so many targets. Before modm we tried xpcc, where the codegen happened during the build step, which was so much worse. Before that we were stuck in macro hell. modm is the result of reducing the vast suffering in our past.

> UsbUart::initialize into a .cpp file.

You can't, since SystemClock is required to be known, so you can [compute prescalers at compile time](https://blog.salkinium.com/computing-baudrates-at-compile-time/).